File: puzzle_3/solution_2.py
import re

sum = 0
enabled = True
with open("input.txt") as f:
    for line in f:
        line = line.strip()
        print(re.split("(don't\\(\\)|do\\(\\))", line))
        for element in re.split("(don't\\(\\)|do\\(\\))", line):
            if element == "don't()":
                enabled = False
            elif element == "do()":
                enabled = True
            elif enabled:
                print(element)
                for match in re.findall("mul\\(([0-9]{1,3}),([0-9]{1,3})\\)", element):
                    sum += int(match[0]) * int(match[1])

print(sum)

File: puzzle_3/solution_2.R
memory = Reduce(paste0, readLines("input.txt"))
valid_instruction = stringr::str_extract_all(paste0("do()",memory,"don't()"), "do\\(\\)(.*?)don't\\(\\)")[[1]]
sum(sapply(valid_instruction, function(instruction) {
    match = stringr::str_match_all(instruction, "mul\\(([0-9]{1,3}),([0-9]{1,3})\\)")[[1]][, 2:3]
    match = matrix(match, ncol=2)
    sum(as.integer(match[,1]) * as.integer(match[,2]))
  }))